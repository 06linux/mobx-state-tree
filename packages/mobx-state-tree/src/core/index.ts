export * from "./node"
export * from "./type"
export { createActionInvoker } from "./action"
export * from "./json-patch"
export * from "./mst-operations"
