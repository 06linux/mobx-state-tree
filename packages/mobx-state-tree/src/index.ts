// Fix some circular deps:
import "./core/type/type"
import "./core/node/object-node"
import "./core/node/scalar-node"

export { types, IModelType, IExtendedObservableMap } from "./types"

export {
    IType,
    ISimpleType,
    IComplexType,
    ISnapshottable,
    typecheckPublic as typecheck
} from "./core/type"

export * from "./core/mst-operations"
export { escapeJsonPath, unescapeJsonPath, IJsonPatch } from "./core/json-patch"
export {
    decorate,
    addMiddleware,
    IMiddlewareEvent,
    IMiddlewareHandler,
    IMiddlewareEventType
} from "./core/action"
export { process } from "./core/process"
export { isStateTreeNode, IStateTreeNode } from "./core/node"

export {
    applyAction,
    onAction,
    IActionRecorder,
    ISerializedActionCall,
    recordActions
} from "./middlewares/on-action"

export { createActionTrackingMiddleware } from "./middlewares/create-action-tracking-middleware"
