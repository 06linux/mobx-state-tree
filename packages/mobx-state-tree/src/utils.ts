import { isObservableArray, $mobx, getAtom } from "mobx"

/** @internal */
declare const global: any

/** @internal */
export const EMPTY_ARRAY: ReadonlyArray<any> = Object.freeze([])
/** @internal */
export const EMPTY_OBJECT: {} = Object.freeze({})

/** @internal */
export const mobxShallow =
    typeof $mobx === "string" ? { deep: false } : { deep: false, proxy: false }
Object.freeze(mobxShallow)

export type IDisposer = () => void

/** @internal */
export function fail(message = "Illegal state"): never {
    throw new Error("[mobx-state-tree] " + message)
}

/** @internal */
export function identity<T>(_: T): T {
    return _
}

/** @internal */
export function nothing(): null {
    return null
}

/** @internal */
export function noop() {}

// pollyfill (for IE) suggested in MDN:
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger
/** @internal */
export const isInteger =
    Number.isInteger ||
    function(value: any) {
        return typeof value === "number" && isFinite(value) && Math.floor(value) === value
    }

/** @internal */
export function isArray(val: any): boolean {
    return !!(Array.isArray(val) || isObservableArray(val)) as boolean
}

/** @internal */
export function asArray<T>(val: undefined | null | T | T[] | ReadonlyArray<T>): T[] {
    if (!val) return (EMPTY_ARRAY as any) as T[]
    if (isArray(val)) return val as T[]
    return [val] as T[]
}

/** @internal */
export function extend<A, B>(a: A, b: B): A & B
/** @internal */
export function extend<A, B, C>(a: A, b: B, c: C): A & B & C
/** @internal */
export function extend<A, B, C, D>(a: A, b: B, c: C, d: D): A & B & C & D
/** @internal */
export function extend(a: any, ...b: any[]): any
/** @internal */
export function extend(a: any, ...b: any[]) {
    for (let i = 0; i < b.length; i++) {
        const current = b[i]
        for (let key in current) a[key] = current[key]
    }
    return a
}

/** @internal */
export function extendKeepGetter<A, B>(a: A, b: B): A & B
/** @internal */
export function extendKeepGetter<A, B, C>(a: A, b: B, c: C): A & B & C
/** @internal */
export function extendKeepGetter<A, B, C, D>(a: A, b: B, c: C, d: D): A & B & C & D
/** @internal */
export function extendKeepGetter(a: any, ...b: any[]): any
/** @internal */
export function extendKeepGetter(a: any, ...b: any[]) {
    for (let i = 0; i < b.length; i++) {
        const current = b[i]
        for (let key in current) {
            const descriptor = Object.getOwnPropertyDescriptor(current, key)!
            if ("get" in descriptor) {
                Object.defineProperty(a, key, { ...descriptor, configurable: true })
                continue
            }
            a[key] = current[key]
        }
    }
    return a
}

/** @internal */
export function isPlainObject(value: any) {
    if (value === null || typeof value !== "object") return false
    const proto = Object.getPrototypeOf(value)
    return proto === Object.prototype || proto === null
}

/** @internal */
export function isMutable(value: any) {
    return (
        value !== null &&
        typeof value === "object" &&
        !(value instanceof Date) &&
        !(value instanceof RegExp)
    )
}

/** @internal */
export function isPrimitive(value: any): boolean {
    if (value === null || value === undefined) return true
    if (
        typeof value === "string" ||
        typeof value === "number" ||
        typeof value === "boolean" ||
        value instanceof Date
    )
        return true
    return false
}

/**
 * @internal
 * Freeze a value and return it (if not in production)
 */
export function freeze<T>(value: T): T {
    if (process.env.NODE_ENV === "production") return value
    return isPrimitive(value) || isObservableArray(value) ? value : Object.freeze(value)
}

/**
 * @internal
 * Recursively freeze a value (if not in production)
 */
export function deepFreeze<T>(value: T): T {
    if (process.env.NODE_ENV === "production") return value
    freeze(value)

    if (isPlainObject(value)) {
        Object.keys(value).forEach(propKey => {
            if (
                !isPrimitive((value as any)[propKey]) &&
                !Object.isFrozen((value as any)[propKey])
            ) {
                deepFreeze((value as any)[propKey])
            }
        })
    }

    return value
}

/** @internal */
export function isSerializable(value: any) {
    return typeof value !== "function"
}

/** @internal */
export function addHiddenFinalProp(object: any, propName: string, value: any) {
    Object.defineProperty(object, propName, {
        enumerable: false,
        writable: false,
        configurable: true,
        value
    })
}

/** @internal */
export function addHiddenWritableProp(object: any, propName: string, value: any) {
    Object.defineProperty(object, propName, {
        enumerable: false,
        writable: true,
        configurable: true,
        value
    })
}

/** @internal */
export function addReadOnlyProp(object: any, propName: string, value: any) {
    Object.defineProperty(object, propName, {
        enumerable: true,
        writable: false,
        configurable: true,
        value
    })
}

/** @internal */
export function remove<T>(collection: T[], item: T) {
    const idx = collection.indexOf(item)
    if (idx !== -1) collection.splice(idx, 1)
}

/** @internal */
export function registerEventHandler(handlers: Function[], handler: Function): IDisposer {
    handlers.push(handler)
    return () => {
        remove(handlers, handler)
    }
}

const prototypeHasOwnProperty = Object.prototype.hasOwnProperty
/** @internal */
export function hasOwnProperty(object: Object, propName: string) {
    return prototypeHasOwnProperty.call(object, propName)
}

/** @internal */
export function argsToArray(args: IArguments): any[] {
    const res = new Array(args.length)
    for (let i = 0; i < args.length; i++) res[i] = args[i]
    return res
}

/** @internal */
export function invalidateComputed(target: any, propName: string) {
    const atom = getAtom(target, propName) as any
    atom.trackAndCompute()
}

/** @internal */
export type DeprecatedFunction = Function & { ids?: { [id: string]: true } }
/** @internal */
export const deprecated: DeprecatedFunction = function(id: string, message: string): void {
    // skip if running production
    if (process.env.NODE_ENV === "production") return
    // warn if hasn't been warned before
    if (deprecated.ids && !deprecated.ids.hasOwnProperty(id)) {
        console.warn("[mobx-state-tree] Deprecation warning: " + message)
    }
    // mark as warned to avoid duplicate warn message
    if (deprecated.ids) deprecated.ids[id] = true
}
deprecated.ids = {}
