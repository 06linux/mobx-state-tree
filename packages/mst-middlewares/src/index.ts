export * from "./atomic"
export * from "./redux"
export * from "./simple-action-logger"
export * from "./time-traveller"
export * from "./undo-manager"
