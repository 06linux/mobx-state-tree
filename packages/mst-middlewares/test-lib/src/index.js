"use strict"
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p]
}
Object.defineProperty(exports, "__esModule", { value: true })
var atomic_1 = require("./atomic")
exports.atomic = atomic_1.default
__export(require("./redux"))
var simple_action_logger_1 = require("./simple-action-logger")
exports.simpleActionLogger = simple_action_logger_1.default
var time_traveller_1 = require("./time-traveller")
exports.TimeTraveller = time_traveller_1.default
var undo_manager_1 = require("./undo-manager")
exports.UndoManager = undo_manager_1.default
