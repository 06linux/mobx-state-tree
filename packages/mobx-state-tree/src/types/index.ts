// tslint:disable-next-line:no_unused-variable
import { IObservableArray, ObservableMap, IAction } from "mobx"
// tslint:disable-next-line:no_unused-variable
import { ISimpleType, IComplexType, IType } from "../core"
// tslint:disable-next-line:no_unused-variable
import { map, IExtendedObservableMap } from "./complex-types/map"
import { array } from "./complex-types/array"
import { identifier } from "./utility-types/identifier"
// tslint:disable-next-line:no_unused-variable
import { model, compose, IModelType } from "./complex-types/model"
import { reference } from "./utility-types/reference"
import { union } from "./utility-types/union"
import { optional } from "./utility-types/optional"
import { literal } from "./utility-types/literal"
import { maybe } from "./utility-types/maybe"
import { refinement } from "./utility-types/refinement"
import { frozen } from "./utility-types/frozen"
import { boolean, DatePrimitive, number, string, undefinedType, nullType } from "./primitives"
import { late } from "./utility-types/late"
import { enumeration } from "./utility-types/enumeration"

export { IModelType, IExtendedObservableMap }

export const types = {
    enumeration,
    model,
    compose,
    reference,
    union,
    optional,
    literal,
    maybe,
    refinement,
    string,
    boolean,
    number,
    Date: DatePrimitive,
    map,
    array,
    frozen,
    identifier,
    late,
    undefined: undefinedType,
    null: nullType
}
