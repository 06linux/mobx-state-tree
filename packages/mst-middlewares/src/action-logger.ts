import * as mst from "mobx-state-tree"

export function actionLogger(call: mst.IMiddlewareEvent, next: any) {
    const skip =
        (call.type === "action" && call.parentId !== 0) ||
        call.type === "flow_resume" ||
        call.type === "flow_resume_error"

    if (!skip)
        console.log(
            `[MST] #${call.rootId} ${call.type} - ${mst.getPath(call.context)}/${call.name}`
        )
    next(call)
}
