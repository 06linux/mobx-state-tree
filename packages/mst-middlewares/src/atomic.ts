import * as mst from "mobx-state-tree"

const atomic = mst.createActionTrackingMiddleware({
    filter: call => call.parentId === 0,
    onStart: call => mst.recordPatches(call.tree),
    onResume: (call, recorder) => recorder.resume(),
    onSuspend: (call, recorder) => recorder.stop(),
    onSuccess: (call, recorder) => {},
    onFail: (call, recorder) => recorder.undo()
})

export default atomic
