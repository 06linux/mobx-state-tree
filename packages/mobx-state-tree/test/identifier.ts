import { types } from "../src"
if (process.env.NODE_ENV !== "production") {
    test("#275 - Identifiers should check refinement", () => {
        const Model = types
            .model("Model", {
                id: types.refinement(
                    "id",
                    types.string,
                    identifier => identifier.indexOf("Model_") === 0
                )
            })
            .actions(self => ({
                setId(id) {
                    self.id = id
                }
            }))
        const ParentModel = types
            .model("ParentModel", {
                models: types.array(Model)
            })
            .actions(self => ({
                addModel(model) {
                    self.models.push(model)
                }
            }))
        expect(() => {
            ParentModel.create({ models: [{ id: "WrongId_1" }] })
        }).toThrow()
        expect(() => {
            const parentStore = ParentModel.create({ models: [] })
            parentStore.addModel({ id: "WrongId_2" })
        }).toThrow()
        expect(() => {
            const model = Model.create({ id: "Model_1" })
            model.setId("WrongId_3")
        }).toThrow()
        expect(() => {
            Model.create({ id: "WrongId_4" })
        }).toThrow()
    })
}
test("#158 - #88 - Identifiers should accept any string character", () => {
    const Todo = types.model("Todo", {
        id: types.identifier(types.string),
        title: types.string
    })
    expect(() => {
        ;["coffee", "cof$fee", "cof|fee", "cof/fee"].forEach(id => {
            Todo.create({
                id: id,
                title: "Get coffee"
            })
        })
    }).not.toThrow()
})
test("#187 - identifiers should not break late types", () => {
    expect(() => {
        const MstNode = types.model("MstNode", {
            value: types.number,
            next: types.maybe(types.late(() => MstNode))
        })
    }).not.toThrow()
})
if (process.env.NODE_ENV !== "production") {
    test("should throw if multiple identifiers provided", () => {
        expect(() => {
            const Model = types.model("Model", {
                id: types.identifier(types.number),
                pk: types.identifier(types.number)
            })
            Model.create({ id: 1, pk: 2 })
        }).toThrowError(
            `[mobx-state-tree] Cannot define property 'pk' as object identifier, property 'id' is already defined as identifier property`
        )
    })
    test("should throw if identifier of wrong type", () => {
        expect(() => {
            const Model = types.model("Model", { id: types.identifier(types.number) })
            Model.create({ id: "1" })
        }).toThrowError(
            `[mobx-state-tree] Error while converting \`{\"id\":\"1\"}\` to \`Model\`:\n\n    at path \"/id\" value \`\"1\"\` is not assignable to type: \`identifier(number)\` (Value is not a number), expected an instance of \`identifier(number)\` or a snapshot like \`identifier(number)\` instead.`
        )
    })
    test("identifier should be used only on model types - no parent provided", () => {
        expect(() => {
            const Model = types.identifier(types.number)
            Model.create(1)
        }).toThrowError(
            `[mobx-state-tree] Identifier types can only be instantiated as direct child of a model type`
        )
    })
}
