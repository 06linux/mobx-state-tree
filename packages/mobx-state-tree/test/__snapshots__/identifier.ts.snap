// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should throw if identifier of wrong type 1`] = `
"[mobx-state-tree] Error while converting \`{\\"id\\":1}\` to \`Model\`:

    at path \\"/id\\" value \`1\` is not assignable to type: \`identifier\` (Value is not a valid identifier, expected a string)."
`;
